src/ProjectOrigin.Vault/Jobs/OnlyRunOncePrReplicaJobBase.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using ProjectOrigin.Vault.Database;
+
+namespace ProjectOrigin.Vault.Jobs;
+
+public abstract class OnlyRunOncePrReplicaJobBase : BackgroundService
+{
+    private readonly string _jobName;
+    private readonly JobKeys _jobKey;
+    private readonly int _runIntervalInSeconds;
+    private readonly IServiceScopeFactory _scopeFactory;
+    private readonly ILogger _logger;
+    private readonly bool _enabled;
+
+    protected OnlyRunOncePrReplicaJobBase(string jobName, JobKeys jobKey, int runIntervalInSeconds, IServiceScopeFactory scopeFactory, ILogger logger, bool enabled)
+    {
+        _jobName = jobName;
+        _jobKey = jobKey;
+        _runIntervalInSeconds = runIntervalInSeconds;
+        _scopeFactory = scopeFactory;
+        _logger = logger;
+        _enabled = enabled;
+    }
+
+    protected abstract Task PerformPeriodicTask(IServiceScope scope, CancellationToken stoppingToken);
+
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        if (!_enabled)
+        {
+            _logger.LogInformation("{jobName} is disabled!", _jobName);
+            return;
+        }
+
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            try
+            {
+                using (var scope = _scopeFactory.CreateScope())
+                {
+                    var unitOfWork = scope.ServiceProvider.GetRequiredService<IUnitOfWork>();
+
+                    var acquiredLock = await unitOfWork.JobExecutionRepository.AcquireAdvisoryLock((int)_jobKey);
```

Comment:
Why is all this locking required?

Reply:
So that only 1 replica does the job, instead of all 3 doing it at the same time